approximate_string_matching/approximate_boyer_moore.py

```diff
@@ -0,0 +1,112 @@
+def boyer_moore_bad_environment(A, p, m, k):
+  bad_table = {(i, a): True for i in range(1, m + 1) for a in A}
+  for i in range(1, m + 1):
+    for j in range(max(0, i - k), min(i + k, m) + 1):
+      bad_table[(j, p[i])] = False
+  return bad_table
+
+
+def boyer_moore_multi_dim_shift(A, p, m, k):
+  ready = {a: m + 1 for a in A}
+  BM_k = {(i, a): m for a in A for i in range(m, m - k - 1, -1)}
+  for i in range(m - 1, 0, -1):
+    for j in range(ready[p[i]] - 1, max(i, m - k) - 1, -1):
+      BM_k[(j, p[i])] = j - i
+    ready[p[i]] = max(i, m - k)
+  return BM_k
+
+
+def approximate_boyer_moore(text, p, n, m, k=None, A=None):
+  if A is None:
+    A = set(list(text[1:] + p[1:]))
+
+  if k is None:
+    k = p.count('?')
+
+  BM_k = boyer_moore_multi_dim_shift(A, p, m, k)
+  BAD = boyer_moore_bad_environment(A, p, m, k)
+
+  j, top = m, min(k + 1, m)
+  D_0 = list(range(m + 1))
+  D_curr = D_0.copy()
```

Comment:
Is it necessary here and L63 to use copy? Why is it the case?

Reply:
Changed to D_0[:]